src/Compilers/CSharp/Portable/FlowAnalysis/VariableUsePass.cs

```diff
@@ -0,0 +1,119 @@
+        /// </summary>
+        /// <remarks>
+        /// An instance of this class is cached on CSharpCompilation, so be careful adding new state.
+        /// (Either avoid state that cannot be shared across the compilation, or do not cache an instance on the compilation)
+        /// </remarks>
+        public VariableUsePass(SourceAssemblySymbol sourceAssembly)
+            : this(sourceAssembly, null, null)
+        {
+        }
+
+        /// <summary>
+        /// Create a visitor for use within a method (e.g. local functions).
+        /// Keep track of what local variables are referenced, by placing them in the HashSets.
+        /// </summary>
+        public VariableUsePass(SourceAssemblySymbol sourceAssembly, HashSet<LocalSymbol> usedVariables, HashSet<LocalFunctionSymbol> usedLocalFunctions)
+        {
+            _sourceAssembly = sourceAssembly;
+            _usedVariablesOpt = usedVariables;
+            _usedLocalFunctionsOpt = usedLocalFunctions;
+        }
+
+        public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+        {
+            _sourceAssembly.NoteFieldAccess(node.FieldSymbol, true, false);
+            return base.VisitFieldAccess(node);
+        }
+
+        public override BoundNode VisitLocal(BoundLocal node)
+        {
+            _usedVariablesOpt?.Add(node.LocalSymbol);
+            return base.VisitLocal(node);
+        }
+
+        private void VisitLocalFunctionReference(LocalFunctionSymbol symbol)
+        {
+            _usedLocalFunctionsOpt?.Add(symbol);
+        }
+
+        public override BoundNode VisitCall(BoundCall node)
+        {
+            if (node.Method.MethodKind == MethodKind.LocalFunction)
+            {
+                var localFunc = (LocalFunctionSymbol)node.Method.OriginalDefinition;
+                VisitLocalFunctionReference(localFunc);
+            }
+
+            return base.VisitCall(node);
+        }
+
+        public override BoundNode VisitConversion(BoundConversion node)
+        {
+            if (node.ConversionKind == ConversionKind.MethodGroup
+                && node.SymbolOpt?.MethodKind == MethodKind.LocalFunction)
+            {
+                var localFunc = (LocalFunctionSymbol)node.SymbolOpt.OriginalDefinition;
+                VisitLocalFunctionReference(localFunc);
+            }
+
+            return base.VisitConversion(node);
+        }
+
+        public override BoundNode VisitDelegateCreationExpression(BoundDelegateCreationExpression node)
+        {
+            if (node.MethodOpt?.MethodKind == MethodKind.LocalFunction)
+            {
+                var localFunc = (LocalFunctionSymbol)node.MethodOpt.OriginalDefinition;
+                VisitLocalFunctionReference(localFunc);
+            }
+
+            return base.VisitDelegateCreationExpression(node);
+        }
+
+        public override BoundNode VisitMethodGroup(BoundMethodGroup node)
+        {
+            foreach (var method in node.Methods)
```

Comment:
What exactly looks suspicious here? If you're discussing the foreach and suggesting only visiting the `[0]`th element, I would argue that looks more suspicious... there's a collection of methods with unknown length, making assumptions that it is length 1 is very strange.